### PR TITLE
Emit CFG in SVG

### PR DESCRIPTION
**Pull request by Cvelth**

### CODE REVIEW

File: include/revng/Yield/Internal/Graph.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
+#pragma once
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Graph.h:1]: This commit says `Outline yield::Graph` but I only see new code. Why is that?

File: lib/Yield/Support/SugiyamaStyleGraphLayout.cpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Yield/Support/SugiyamaStyleGraphLayout.h"
+
+bool yield::sugiyama::layout(Graph &Graph, const Configuration &Configuration) {
+}
```
Comment (aleclearmind) [inline lib/Yield/Support/SugiyamaStyleGraphLayout.cpp:2326]: Uh?

**Review by aleclearmind — CHANGES_REQUESTED** [review]
Sugiyama is gigantic, directory + multiple files? You seem to already have split things nicely into commits.

File: lib/Yield/Internal/Verify.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Yield/Internal/BasicBlock.h"
+#include "revng/Yield/Internal/Function.h"
+#include "revng/Yield/Internal/Instruction.h"
+#include "revng/Yield/Internal/Tag.h"
+
+bool yield::Tag::verify(model::VerifyHelper &VH) const {
+  VH.maybeFail(Type != TagType::Invalid);
```
[the remark this answers is not in the document]
Comment (Cvelth) [reply lib/Yield/Internal/Verify.cpp:14]: 1. Yeah, I had no idea how that part of the model worked so I just copied what I saw. I'll fix this :D

2. Love the idea!

File: lib/Pipes/Yield/AssemblyPipes.cpp [hunk 4]
```diff
@@ -44,20 +45,19 @@ void YieldAssemblyPipe::run(pipeline::Context &Context,
...
 }
 
-void YieldAssemblyPipe::print(const pipeline::Context &,
-                              llvm::raw_ostream &OS,
-                              llvm::ArrayRef<std::string> Files) const {
+void ProcessAssemblyPipe::print(const pipeline::Context &,
+                                llvm::raw_ostream &OS,
+                                llvm::ArrayRef<std::string> Files) const {
   OS << *revng::ResourceFinder.findFile("bin/revng") << " yield assembly html "
      << Files[0] << " " << Files[1] << '\n';
```
[the remark this answers is not in the document]
Comment (Cvelth) [reply lib/Pipes/Yield/AssemblyPipes.cpp:56]: The tools weren't changed, so it is as correct as it is in the current develop. But the tools still need a little bit of love. I think these shouldn't be in the `yield` "~~name~~-tool-space" after all - they are way too internal-like. I can't really imagine user wanting to do much with these. I imagine the python script I've been using for simple binary-to-a-directory-with-htmls have should become the foundation on `revng yield` instead.

File: include/revng/Yield/ControlFlow/Configuration.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+  Graph::Dimension InstructionFontSize;
+  Graph::Dimension InstructionAddressFontSize;
+  Graph::Dimension InstructionBytesFontSize;
+  Graph::Dimension CommentFontSize;
+  Graph::Dimension LabelFontSize;
+
+  size_t NodeCornerRoundingFactor;
+
+  bool UseOrtogonalBends;
```
[the remark this answers is not in the document]
Comment (Cvelth) [reply include/revng/Yield/ControlFlow/Configuration.h:30]: Whoops :D

File: include/revng/Yield/Internal/Graph.h (hunk 1 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Graph.h:1]: It only OUTLINEs an interface. The actual usage of the new interface is introduced in the next commit.

File: lib/Yield/Support/SugiyamaStyleGraphLayout.cpp (hunk 2 above)
Comment (Cvelth) [reply lib/Yield/Support/SugiyamaStyleGraphLayout.cpp:2326]: This commit (similarly to the previous thread) only describes the interface. The actual functionality comes in the next commits.

File: lib/Yield/Assembly/HTML.cpp [hunk 6]
```diff
@@ -527,62 +637,120 @@ static std::string instruction(const assembly::Instruction &Instruction,
...
+    return llvm::formatv(templates::SimpleDiv,
+                         tags::BasicBlock,
+                         std::move(Result));
+  else
+    return llvm::formatv(templates::BlockDiv,
+                         tags::BasicBlock,
+                         basicBlockID(BasicBlock.Address),
+                         std::move(Result));
+}
+#include <iostream>
```
Comment (aleclearmind) [inline lib/Yield/Assembly/HTML.cpp:693]: Wat

Comment (Cvelth) [reply lib/Yield/Assembly/HTML.cpp:693]: 🤦🏻, an indication of some painful debugging at one point I forgot to remove. There's little as embarrassing as for someone to see your dirty includes 😓

File: lib/Yield/Internal/Verify.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Cvelth) [reply lib/Yield/Internal/Verify.cpp]: 1. Done.
2. I'm closing this thread, but hopefully this idea doesn't get lost.

File: include/revng/Yield/ControlFlow/Configuration.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Cvelth) [reply include/revng/Yield/ControlFlow/Configuration.h]: Fixed.

File: lib/Yield/Assembly/HTML.cpp (hunk 6 above)
Comment (Cvelth) [reply lib/Yield/Assembly/HTML.cpp]: As if it was never there.

File: lib/Pipes/Yield/AssemblyPipes.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Cvelth) [reply lib/Pipes/Yield/AssemblyPipes.cpp]: Dropped the tools: https://github.com/revng/revng/pull/179/commits/742dbf8e03c4201f48bb5bce3798cde51b090e86.

File: include/revng/Yield/Internal/Graph.h (hunk 1 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Graph.h:1]: Uhm OK, as long as it compiles.

File: lib/Yield/Support/SugiyamaStyleGraphLayout.cpp (hunk 2 above)
Comment (aleclearmind) [reply lib/Yield/Support/SugiyamaStyleGraphLayout.cpp]: At least put a `revng_abort`. Doesn't this trigger warnings?

File: include/revng/Yield/Internal/Graph.h (hunk 1 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Graph.h:1]: It does, that's why the empty bodies and stuff are there.

File: lib/Yield/Support/SugiyamaStyleGraphLayout.cpp (hunk 2 above)
Comment (Cvelth) [reply lib/Yield/Support/SugiyamaStyleGraphLayout.cpp]: `orc install revng@debug` didn't complain, so I guess not.

`revng_abort` works.

File: include/revng/Yield/Internal/Graph.h (hunk 1 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Graph.h:1]: Closing this one, keeping the second one open for now.

File: lib/Yield/Support/SugiyamaStyleGraphLayout.cpp (hunk 2 above)
Comment (Cvelth) [reply lib/Yield/Support/SugiyamaStyleGraphLayout.cpp]: Done.

File: include/revng/Yield/Internal/Function.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+/* TUPLE-TREE-YAML
+
+name: Function
+type: struct
+fields:
+  - name: Entry
+    type: MetaAddress
+  - name: Name
+    type: std::string
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Function.h:24]: If possible, we should avoid copying names in intermediate artifacts, since it needlessly increase the situations under which we need to recompute this.

tl;dr Can you fetch this from the model later down the pipeline?

File: include/revng/Yield/Internal/Function.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  - name: BasicBlocks
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Function.h:26]: Would it make sense to
```suggestion
  - name: ControlFlowGraph
```
For coherence with EFA?

File: include/revng/Yield/Internal/BasicBlock.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      upcastable: true
+      elementType: efa::FunctionEdgeBase
+
+  - name: Instructions
+    sequence:
+      type: SortedVector
+      elementType: yield::Instruction
+
+  - name: IsLabelAlwaysRequired
+    type: bool
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/BasicBlock.h:45]: Doc

File: include/revng/Yield/Internal/BasicBlock.h [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+  - name: CommentIndicator
+    type: std::string
+  - name: LabelIndicator
+    type: std::string
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/BasicBlock.h:50]: * [ ] What are these used for? Can you add doc?
* [ ] Drop `Indicator` suffix

File: include/revng/Yield/Internal/Instruction.h [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+using ByteContainer = llvm::SmallVector<uint8_t, 16>;
+}
+
+/* TUPLE-TREE-YAML
+
+name: Instruction
+type: struct
+fields:
+  - name: Address
+    type: MetaAddress
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Instruction.h:26]: Can we have `Start`/`End`?

File: include/revng/Yield/Internal/Instruction.h [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  - name: Bytes
+    type: yield::ByteContainer
+  - name: Raw
+    type: std::string
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Instruction.h:30]:
```suggestion
  - name: RawBytes
    type: yield::ByteContainer
  - name: Disassembly
    type: std::string
```

File: include/revng/Yield/Internal/Instruction.h [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+  - name: Tags
+    sequence:
+      type: std::vector
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Instruction.h:34]: * Consider using `SortedVector` and make `yield::Tag` with a key spanning all the fields of the data structure? It can work as a set, avoiding issues with duplicate tags.
* I'm afraid `mnemonic` +  tag set will inflate in a more and more sophisticated data structure and I really wonder if we should avoid representing the structure here entirely and just switch to XML at this stage, i.e., drop mnemonic and tags and stuff XML into the `Raw`/`Disassembly` field. We're going from 1) pseudo-XML from LLVM disassembler, 2) to out-of-line markup here, 3) back to XML when we'll ship this to the client and finally 4) in VSCode we'll translate it back to out-of-line markup again. Maybe we can avoid 2) and go from 1) to 3)?

File: include/revng/Yield/Internal/Instruction.h [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+      elementType: yield::Tag
+    optional: true
+
+  - name: Opcode
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Instruction.h:38]: * Uhm, I feel like *opcode* identifies the number identifying the instruction, while here probably the most appropriate term is *mnemonic*, I think.
* Why do we need this?

File: include/revng/Yield/Internal/Instruction.h [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    type: std::string
+    optional: true
+  - name: Comment
+    type: std::string
+    optional: true
+  - name: Error
+    type: std::string
+    optional: true
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Instruction.h:46]: What is this for? Doc.

File: include/revng/Yield/Internal/Instruction.h [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+  - name: HasDelayedSlot
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Instruction.h:48]: What is the semantic of this field? Is this `yield::Instruction` representing both the jump instruction and the instruction in the delay slot?

```suggestion
  - name: HasDelaySlot
```

File: include/revng/Yield/Internal/Tag.h [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+name: Tag
+type: struct
+fields:
+  - name: Type
+    type: yield::TagType::Values
+  - name: FromPosition
+    type: size_t
+  - name: ToPosition
+    type: size_t
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/Tag.h:23]: Please use portable types in serializable data structures.
```suggestion
  - name: From
    type: uint64_t
  - name: To
    type: uint64_t
```

File: include/revng/Yield/Internal/BasicBlock.h [hunk 18]
```diff
@@ -0,0 +1,79 @@
+#pragma once
```
Comment (aleclearmind) [inline include/revng/Yield/Internal/BasicBlock.h:1]: Can you find a more appropriate name than `Internal`? This seems like a relic from caliban.

File: include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+template<IsGenericGraph GraphType,
+         IsBasicBlock BasicBlockType,
+         typename ...OtherTs,
+         template<typename...> typename Container>
+  requires std::is_constructible_v<typename GraphType::Node,
+                                   const MetaAddress &>
+std::pair<GraphType, std::map<MetaAddress, typename GraphType::Node *>>
+buildControlFlowGraph(const Container<BasicBlockType, OtherTs...> &BasicBlocks,
+                      const MetaAddress &EntryAddress,
+                      const model::Binary &Binary) {
```
Comment (aleclearmind) [inline include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h:81]: See if you can make this more accessible and add some doc.

File: lib/Yield/Internal/Verify.cpp [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+bool yield::Instruction::verify(model::VerifyHelper &VH) const {
+  if (Address.isInvalid())
+    return VH.fail("An instruction has to have a valid address.");
+  if (Raw.empty())
+    return VH.fail("A raw view of an instruction cannot be empty.");
+  if (Bytes.empty())
+    return VH.fail("An instruction has to be at least one byte big.");
+
+  // TODO: Tags might need special verification as well.
+  // Their order, the fact that they don't overlap in undesirable ways, etc.
```
Comment (aleclearmind) [inline lib/Yield/Internal/Verify.cpp:35]: If you switch to a `SortedVector` of elements sorted by `Start, End, Type` you have no problems with ordering by construction!

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: include/revng/Pipes/FunctionStringMap.h [hunk 21]
```diff
@@ -23,6 +25,7 @@ class FunctionStringMap : public pipeline::Container<FunctionStringMap> {
   struct String {
     std::string TheString;
     operator std::string() const { return TheString; }
+    operator llvm::StringRef() const { return TheString; }
```
[the remark this answers is not in the document]
Comment (Cvelth) [reply include/revng/Pipes/FunctionStringMap.h:28]: This causes problems, sadly, I already have an idea on how to solve it, once it's a bit more develop, I'll ping `@fez` and discuss it with him.

File: include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h (hunk 19 above)
Comment (Cvelth) [reply include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h:81]: Always a good idea! :)

File: include/revng/Yield/Internal/Tag.h (hunk 17 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Tag.h:23]: Ah, right, sorry.

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:48]: This field is set to true for an instructions that contains a delayed slot. The actual delayed instruction is an instruction that follows the one with this field set.

I went with this approach because that's how MCDissassembler stores these. This allows generate instructions independently from each other - no need to carry over any information between neighbouring ones.

File: include/revng/Yield/Internal/Instruction.h (hunk 15 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:46]: Sure, sorry.

File: include/revng/Yield/Internal/Instruction.h (hunk 13 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:34]: I feel like this way to store it gives the most flexibility possible. If at some point we decide to change the approach, there'll be a lot less work to do with having tags and the raw text separated.

File: include/revng/Yield/Internal/Instruction.h (hunk 12 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:30]: Sure.

File: include/revng/Yield/Internal/Instruction.h (hunk 11 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:26]: Sure.

Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:26]: Do we really need `End` though? Consider that instructions are always stored in a vector and are considered packed. As such the `End` of one instructions is just the `Start` of the next one. The `End` of the last instruction in a basic block is the `End` of the corresponding block.

File: include/revng/Yield/Internal/BasicBlock.h (hunk 10 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/BasicBlock.h:50]: These are what indicates `Comment` and `Label` correspondingly in the official way to print assembly. For example, for most platforms `LabelIndicator` is `:`.

`CommentIndicator` varies wildly, like `@` for arm, `//` for aarch64 and so on.

2. Do you have better name ideas then? I feel like just calling them `Comment` and `Label` is going to get really confusing.

File: include/revng/Yield/Internal/Function.h (hunk 7 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Function.h:24]: The only issue with that is that it'd need to depend on the model in the later stages as well (as of now they are self-contained, as in they don't need anything but this intermediate structure).

But, now that you say it, yeah that would make a lot of sense! There's no reason to invalidate the entirety of this intermediate structure just because user renamed a function.

What's the safest way to store a reference to a model function? `TupleReference`? Are there problems if I were to serialize those outside the model? As in, model could change unexpectedly between serializing and deserializing those.

File: include/revng/Yield/Internal/BasicBlock.h (hunk 18 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/BasicBlock.h:1]: To me it makes sense to have the file that defines `yield::BasicBlock` in `include/revng/Yield/BasicBlock.h`.
Those are data structures as valid as EFA's ones.
All our data structures are always WIP, I wouldn't worry too much.

File: lib/Yield/Internal/Verify.cpp (hunk 20 above)
Comment (Cvelth) [reply lib/Yield/Internal/Verify.cpp:35]: I actually not sure whether we need that or not. I can't think of a case where there are two tags beginning and ending at the same points.

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h:48]: OK, document this. Unless you saw *delayed* somewhere let's go with *delay*.

Also, in the verify assert that this is the semi-last instruction of the list.

File: include/revng/Yield/Internal/BasicBlock.h (hunk 18 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/BasicBlock.h:1]: Yeah, thinking about it, I'm not opposed to dropping the `Internal` "namespace".

File: include/revng/Yield/Internal/Instruction.h (hunk 14 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h:38]: So this is the name *LLVM* gives to the opcode? Then consider calling it `LLVMOpcode`.
Also, consider that this data structure could be reused with different disassemblers one day.

File: include/revng/Yield/Internal/Instruction.h (hunk 11 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h:26]: Works, but then state this somewhere in the doc and check this in verify.

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:48]: I looked it up, it's my bad. For some reason I mixed up `delayed load` and `delay slot` when naming this. Will fix.

2. Do we care where it is really? Sure, we should probably check that it's not the last, but other than that let the instruction be wherever it is, imo.

File: include/revng/Yield/Internal/Instruction.h (hunk 14 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:38]: 1. yeah, it's llvm notation, sorry for not making that clear.

2. That's actually the reason why I kept it as just an `opcode`, every disassembler should theoretically have something similar after all - hence LLVM in the name could be redundant.

File: include/revng/Yield/Internal/Function.h (hunk 7 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Function.h:24]: I feel like not depending on the model is a bit of a misguided goal.
The safest thing is to use the `Entry` address to make a lookup on the model and get the Right Name™.

File: lib/Yield/Internal/Verify.cpp (hunk 20 above)
Comment (aleclearmind) [reply lib/Yield/Internal/Verify.cpp:35]: ACK, the  consider implementing a custom `operator<`.

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h:48]: 1. :+1:
2. Mmh, I like our data structures to offer strong, useful, predictable guarantees. If in our mind we expect something, why not enforce it?

File: include/revng/Yield/Internal/BasicBlock.h (hunk 10 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/BasicBlock.h:50]: Yeah, that was my main worry as well, but keeping track of those manually didn't felt right. This allows as to just LLVM for these, but it also means we store a whole lot of duplicates.

I thought about moving them higher up, but the highest we can possible have them in the current datastructure state is at a function level which isn't much better than the basic block level.

Do you prefer we store these in the headers somewhere then?

File: include/revng/Yield/Internal/Function.h (hunk 7 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Function.h:24]: Feels like that should work.

File: lib/Yield/Internal/Verify.cpp (hunk 20 above)
Comment (Cvelth) [reply lib/Yield/Internal/Verify.cpp:35]: What a weird way to write "the spaceship operator" /s

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:48]: 2. Mostly, in this case, if we try to enforce something this rigid, I see no point of having this flag at an instruction level. If the delay slot can only be located at the very end of a block, it should be a block-level flag!

Now that I think about it, maybe that's actually the way to go. What do you think? Should `EndsInDelaySlot` be a basic-block field after all?

File: include/revng/Yield/Internal/BasicBlock.h (hunk 10 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/BasicBlock.h:50]: In [`Architecture.h` we already have similar stuff](https://github.com/revng/revng/blob/develop/include/revng/Model/Architecture.h#L258).

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h:48]: Yes!

File: include/revng/Yield/Internal/Instruction.h (hunk 14 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h:38]: `OpcodeIdentifier`? We can decouple this from LLVM and just provide the guarantee that the identifier uniquely identifies the instruction in this architecture.
If you prefer, not to change the name, make sure the doc explains something as the thing above.

File: include/revng/Yield/Internal/Instruction.h (hunk 13 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h:34]: OK, let's stick with this.

I'm closing this discussion in favor of the other one where you mentioned `<=>`.

File: include/revng/Yield/Internal/Instruction.h (hunk 14 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h:38]: Sure, I don't really have any attachment to the name.

File: include/revng/Yield/Internal/BasicBlock.h (hunk 9 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/BasicBlock.h]: [Done](https://github.com/Cvelth/revng/blob/snapshot/wip_cfg_in_svg/include/revng/Yield/BasicBlock.h#L44).

File: include/revng/Yield/Internal/Function.h (hunk 8 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Function.h]: [Done](https://github.com/Cvelth/revng/blob/snapshot/wip_cfg_in_svg/include/revng/Yield/Function.h#L24).

File: include/revng/Yield/Internal/BasicBlock.h (hunk 10 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/BasicBlock.h]: Done: https://github.com/Cvelth/revng/blob/snapshot/wip_cfg_in_svg/include/revng/Model/Architecture.h#L366.

File: include/revng/Yield/Internal/Instruction.h (hunk 11 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h]: [Done](https://github.com/Cvelth/revng/blob/snapshot/wip_cfg_in_svg/lib/Yield/Verify.cpp#L53).

File: include/revng/Yield/Internal/Instruction.h (hunk 12 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h]: [Done](https://github.com/Cvelth/revng/blob/snapshot/wip_cfg_in_svg/include/revng/Yield/Instruction.h#L29).

File: include/revng/Yield/Internal/Instruction.h (hunk 14 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h]: [Done](https://github.com/Cvelth/revng/blob/snapshot/wip_cfg_in_svg/include/revng/Yield/Instruction.h#L40).

File: include/revng/Yield/Internal/Instruction.h (hunk 15 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h]: [Done](https://github.com/Cvelth/revng/blob/snapshot/wip_cfg_in_svg/include/revng/Yield/Instruction.h#L50).

File: include/revng/Yield/Internal/Tag.h (hunk 17 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Tag.h]: [Done](https://github.com/Cvelth/revng/blob/snapshot/wip_cfg_in_svg/include/revng/Yield/Tag.h#L20).

File: include/revng/Yield/Internal/BasicBlock.h (hunk 18 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/BasicBlock.h]: [Done](https://github.com/Cvelth/revng/tree/snapshot/wip_cfg_in_svg/include/revng/Yield).

File: include/revng/Pipes/FunctionStringMap.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (Cvelth) [reply include/revng/Pipes/FunctionStringMap.h]: I went like this instead: https://github.com/Cvelth/revng/commit/b75fbf4fcb4ce8250d46443ecf5b41ccfb26ae94

What do you think?

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h]: What do you think?
https://github.com/revng/revng/pull/179/commits/c514425b189fe2fcbf337ff546726575e8c3c0f6

File: include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h (hunk 19 above)
Comment (Cvelth) [reply include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h:100]: Something like this? https://github.com/revng/revng/pull/179/commits/17c6a939a0f08d8e81eb151378dbadbb6ef878d9

File: lib/Yield/Internal/Verify.cpp (hunk 20 above)
Comment (Cvelth) [reply lib/Yield/Internal/Verify.cpp]: What do you think?
https://github.com/revng/revng/pull/179/commits/a9a153ef846bdce867a44a0baa2ced49f31ce148 and https://github.com/revng/revng/pull/179/commits/d03f75b9360e0862c498787af8908a1d304c2de1

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h]: Looks good!

* I'd go for `HasDelaySlot`, there's doc for better specifying the meaning.
* `InstructionReturnValueTuple`: you often do this, can you find a more semantic name instead of simply describing the role? I find that providing more dignity to data structure is always beneficial.

File: include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h (hunk 19 above)
Comment (aleclearmind) [reply include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h:100]: Yeah!

File: include/revng/Yield/Internal/Instruction.h (hunk 16 above)
Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h]: 1. Sure.
2. I do it for the data structures I wouldn't have added if I could get away with it. For example, in this case it's literally `std::tuple<yield::Instruction, bool, size_t>` except that's too long and somewhat confusing (What does this `bool` mean? What's the int for?), hence the struct with named fields instead. If you have any name suggestions - I'm not opposed, BUT, it's only ever used as the return value of this specific function and is expected to be unrolled using structured bindings at the call site.

Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h]: 2. It wouldn't provide it much dignity to call it its real name: `InstructionAndTheFieldsThatWereRemovedFromIt` :0

Comment (aleclearmind) [reply include/revng/Yield/Internal/Instruction.h]: I'd call it "LLVMDisassemblerInterface::Instruction" and be done with it. Its scoped, no one will use it for other purpose, it makes sense as a name. Occam's razor!

Comment (Cvelth) [reply include/revng/Yield/Internal/Instruction.h]: I guess that's one way to do it. It is somewhat confusing to have `Instruction` class that has an `Instruction` member with type `Instruction` but, either way, this is not much worse the the current solution, so I'm fine with that.
